### PR TITLE
Ar tag detection 481

### LINKED ISSUE #481: Implement AR tag detection
Based on the code from ERC https://github.com/space-concordia-robotics/docker-erc

This task would require to have a node with the responsibilities to

- Take a stream as an input and republish it with AR tag information
- The node should be able to handle more than one stream (we have 3 cameras on the rover). Might be hard to test with only 1 webcam though
- If possible, the AR tag detection should actually track the ar tags on the screen, but just any information indicating there's an AR tag on the screen should be good enough
- The node should be launched in the appropriate launch files
- The AR tag affected streams should be visible in the GUI

### PR DESCRIPTION
**Pull request by MewtR**
# Assignee Section

## Description
This pull request integrates the ar tag detection from [here](https://github.com/space-concordia-robotics/docker-erc) to this repository.

#### Prerequisites
- Make sure you install [ar_track_alvar](http://wiki.ros.org/ar_track_alvar): `sudo apt install ros-melodic-ar-track-alvar`
- Calibrate your camera by following the instructions [here](http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration)
- Print out some AR tags. They can be found [here](http://wiki.ros.org/ar_track_alvar?action=AttachFile&do=view&target=markers0to8.png) for example.

### Steps for Testing
1. Run the project and visit `http://localhost:5000`.
2. Select a camera stream and start it.
3. Click on the "AR" button, you should see a red square appear.
4. Hold up the printed AR tags in front of you camera. The red square should turn green.
5. Click on the "AR" button again and the square should disappear.

#### Additional Steps
1. Test with multiple streams simultaneously.

closes #481

The approval from all software team leads is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [ ] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

### CODE REVIEW

File: robot/basestation/static/js/elements/feed/camera.js [hunk 1]
```diff
@@ -135,7 +138,6 @@ $(document).ready(() => {
     cameraPower.attr('src', POWER_ON)
 
     cameraControls.attr('camera-name', cameraName)
-    console.log("memes")
```
Comment (MartensCedric) [inline robot/basestation/static/js/elements/feed/camera.js:138]: lmfao I'm dead @RealWilliamWells @PeterGhimself @vashmata @AFknUsername

Comment (RealWilliamWells) [reply robot/basestation/static/js/elements/feed/camera.js:138]: lol, ngl that was prob me. I always print "memes" if I'm using console.log to test

File: robot/rospackages/src/task_handler/scripts/Listener.py [hunk 2]
```diff
@@ -77,8 +77,10 @@ def stop(self, p_keyword=''):
                         real_pid = get_pid("cv_camera_node")
 
                     if real_pid != -1:
-                        output, error = run_shell("kill -9", str(real_pid))
-                        print("kill -9", str(real_pid))
+                        # better to use SIGINT rather than SIGKILL because it kills all
+                        # child processes as well
```
Comment (RealWilliamWells) [inline robot/rospackages/src/task_handler/scripts/Listener.py:81]: Nice, good catch

**Review by RealWilliamWells — CHANGES_REQUESTED** [review]
Good work gamer. Just a couple things, you should add the installation of ros-melodic-ar-track-alvar to the environment setup script and the dockerfile. 

Another thing too is that we may want to reduce unnecessary bandwidth usage by having something in place to limit it to one AR stream at a time. To avoid having essentially copies of every stream being published from the rover. Maybe this might not be a big concern tho or maybe there's a better solution. What do you think gamers? @MartensCedric  @AFknUsername

**Comment by MewtR** [comment]
@RealWilliamWells 
Now ros-track-alvar is installed through the scripts and the Dockerfile (although for docker it might require a bit more work to get right because as far as I know you have to specify the device name you want mounted inside your container. In other words if it isn't named `/dev/video0`, it won't work.)

But yeah you're right performance is pretty trash at the moment. I'll see if I can make it better.

File: robot/rospackages/src/camera_vision/scripts/ar_tracker.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import sys
+
+import rospy
+import cv2
+from sensor_msgs.msg import Image, CameraInfo
+from ar_track_alvar_msgs.msg import AlvarMarkers
+from cv_bridge import CvBridge, CvBridgeError
+import numpy as np
+
+class arTrackerDemo():
```
Comment (MartensCedric) [inline robot/rospackages/src/camera_vision/scripts/ar_tracker.py:11]: Does it need to be called Demo

**Review by MartensCedric — CHANGES_REQUESTED** [review]

File: robot/rospackages/src/camera_vision/scripts/ar_tracker.py (hunk 3 above)
Comment (MewtR) [reply robot/rospackages/src/camera_vision/scripts/ar_tracker.py]: No it does not. Changed it to `arTracker`

File: robot/rospackages/src/webcam_ar_track/package.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  <!-- One maintainer tag required, multiple allowed, one person per tag -->
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="astro@todo.todo">astro</maintainer>
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Comment (VimFreeman) [inline robot/rospackages/src/webcam_ar_track/package.xml:16]: Please complete TODOs before merging

**Review by VimFreeman — CHANGES_REQUESTED** [review]
Please don't leave TODOs in PRs
Otherwise, nice clean documented code :+1:
Will test and report back soon

File: robot/rospackages/src/webcam_ar_track/package.xml (hunk 4 above)
Comment (MewtR) [reply robot/rospackages/src/webcam_ar_track/package.xml]: Done

**Comment by MartensCedric** [comment]
FYI the name of the cameras on the rover are 

videoArmScience
videoFront
videoRear

as shown in the symlinks created here https://github.com/space-concordia-robotics/robotics-prototype/blob/master/robot/rover/config/udev/rules.d/my-webcams.rules

**Review by MartensCedric — CHANGES_REQUESTED** [review]
task_handler server crashes on this PR (not in master) going to look into it. 

Here's the crash : 
image

**Comment by MartensCedric** [comment]
We have trouble with setting up the configuration for the cameras. It seg faults. From the source code  : 

It saves a tar file in /tmp, this works, but what does not work is the service proxy step

http://wiki.ros.org/rospy/Overview/Services

**Comment by MartensCedric** [comment]
Could you add a comment in the .launch files about the arguments needed in kinetic but not melodic (just so if someone struggles to get this working on kinetic for whatever reason they wont go bonkers)

**Comment by MewtR** [comment]
Closed in favor of #492

**Pull request closed without merging**